Max weighted matching

jpweytjens opened the pull request:
This pull request implements max weighted bipartite graph matching as a method for the `OneToOne` class. It uses `networkx` for the actual matching algorithm. Multiple helper functions are defined to transform the `MultiIndex` output from `recordlinkage` to a  `graph` object and back.

jpweytjens commented:
You can easily visualize (small) graphs with the following code

```
def draw_weighted_bipartite_graph(graph):
    """Draw a weighted bipartite graph. 
    No garantuees are made about the order of the nodes."""

    labels = nx.get_edge_attributes(graph, "weight")
    nodes = [n for n, d in graph.nodes(data=True) if d["bipartite"] == 0]
    pos = nx.drawing.layout.bipartite_layout(graph, nodes)

    nx.draw_networkx_edge_labels(graph, pos, labels)
    nx.draw_networkx(graph, pos)
```

though I'm not sure `recordlinkage` should contain such a function?

jpweytjens commented:
If I find the time, I'm interesting in adding the symmetric best matching and stable matching methods from [Franke et. al](https://dbs.uni-leipzig.de/file/main.pdf) as well. These methods seem to produce comparable results, for a much lower computational cost.

J535D165 commented:
Thanks @jpweytjens 

> You can easily visualize (small) graphs with the following code
> 
> ```
> def draw_weighted_bipartite_graph(graph):
>     """Draw a weighted bipartite graph. 
>     No garantuees are made about the order of the nodes."""
> 
>     labels = nx.get_edge_attributes(graph, "weight")
>     nodes = [n for n, d in graph.nodes(data=True) if d["bipartite"] == 0]
>     pos = nx.drawing.layout.bipartite_layout(graph, nodes)
> 
>     nx.draw_networkx_edge_labels(graph, pos, labels)
>     nx.draw_networkx(graph, pos)
> ```
> 
> though I'm not sure `recordlinkage` should contain such a function?

That would be nice to have in the documentation or on the examples page. If it turns out to be a succes, we add it to the toolkit. 

Documentation: https://recordlinkage.readthedocs.io/en/latest/ref-classifiers.html?highlight=onetoone#network

jpweytjens commented on recordlinkage/network.py:
```diff
@@ -50,27 +66,136 @@ def _compute_greedy(self, links):
+
+        return multiindex
+
+    def compute(self, links, data=None):
```
I agree that passing a `Series` would be better. I'm not sure what the best way would be to create this `Series`.

This commit sums the similarity weights determined with the `Compare` class. Another, perhaps better option, would be to use the match probabilities as calculated by `prob()` method from the fitted classifier. These two methods require different inputs though, the calculated features for the former, the fitted classifier for the latter. 

Should the `OneToOneLinking` class provide a method to make this series that can handle both methods? Perhaps `utils` is a better place for this function? 

What do you think is more clear for the user?

jpweytjens commented:
I've added a new commit that should address all the requested changes.

- A new function `add_weights` is introduced to add weights to candidate matches with 3 different methods. This function returns a Series instead of DataFrame.
- Series are now used throught the `max_weighted` matching function, instead of DataFrames.
- networkx has been removed from tox and added to `setup.py`
- Documentation has been added for the `max_weighted` method and an example is given for the new `add_weights` function.

jpweytjens commented:
Can you review the changes @J535D165 ?

jpweytjens commented:
> Can you review the changes @J535D165 ?

Any update on this PR @J535D165 ?

jpweytjens commented:
Can you review the changes @J535D165 ?
